pkg/lib/v0_2_0/events.go

```diff
@@ -65,14 +66,24 @@ func NewHTTPEventSender(endpoint string) (*HTTPEventSender, error) {
 	return httpSender, nil
 }
 
-// SendEvent sends a CloudEvent
+// Deprecated: use HTTPEventSender.Send instead
 func (httpSender HTTPEventSender) SendEvent(event cloudevents.Event) error {
 	ctx := cloudevents.ContextWithTarget(context.Background(), httpSender.EventsEndpoint)
 	ctx = cloudevents.WithEncodingStructured(ctx)
 	return httpSender.Send(ctx, event)
 }
 
 func (httpSender HTTPEventSender) Send(ctx context.Context, event cloudevents.Event) error {
+	// if not already added by the caller, add the one from the Sender
+	if cloudevents.TargetFromContext(ctx) == nil {
+		ctx = cloudevents.ContextWithTarget(ctx, httpSender.EventsEndpoint)
+	}
+
+	// TODO: Should we also add cloudevents.WithEncodingStructured(ctx) here to avoid duplication?
+	// There doesn't seem to be a way to check if already present in the ctx, as the key
```

Comment:
Could you please elaborate more about this duplication problem? it is not clear to me

Reply:
Sure. So far, several components are using the `SendEvent` method, like those: https://grep.app/search?q=SendEvent%28&filter[repo][0]=keptn/keptn. The `SendEvent` method did these two things inside: 

```go
ctx := cloudevents.ContextWithTarget(context.Background(), httpSender.EventsEndpoint)
ctx = cloudevents.WithEncodingStructured(ctx)
```

To properly get context propagation, code should use the `Send` one, which accepts the `context`. But that method does not do these two steps. In my change I was able to check if the `ctx` already has the "target" in it, but I did not find a way to check if it already had the "WithEncodingStructured" in it. 

This means that when changing code that uses `SendEvent` to use `Send`, we'll have to do `ctx = cloudevents.WithEncodingStructured(ctx)` again and again, thus the duplication. 

I might be wrong, but it seems the point of the `Send` is to be "agnostic" of the encoding/target. It makes it flexible for callers to encode with whatever they want and send to maybe a different receiver, than the one configured in the `EventSender`. 

So we are either OK of adding `ctx = cloudevents.WithEncodingStructured(ctx)` where we previously didn't have to, OR if we know that we will only be working with Structured encoding, then it might be OK to just add to `Send` as well. 

Did that clarify it?

Reply:
In Keptn we only support events that have been sent with Structured encoding, so adding it to `Send` is fine in my opinion

Reply:
Alright! I changed then to use structured encoding by default. Also added tests to ensure if it is changed later on we'll catch it.